### PR TITLE
fix(compile): survive binary/workspace skew; zstd, DOMException-subclass and panic-dedup gaps hit compiling a real npm CLI

### PR DESCRIPTION
**Pull request by jdalton**
Compiling [Socket Firewall](https://github.com/SocketDev/firewall) (`sfw` — a TLS-MITM package-manager proxy with undici, node-forge, iovalkey, zod, lodash, … in its module graph) end-to-end surfaced four independent blockers. Each is fixed at its own layer; together they take sfw from 'fails at dependency collection' to a running native arm64 binary.

## 1. Auto-optimize feature skew → silent doomed link
The `perry` binary's baked-in cross-feature list tracks the branch the binary was **built** from; the auto-optimize cargo build resolves it against the **checkout on disk**. One `perry-runtime/<feat>` the checkout doesn't declare fails the whole cargo resolve, and the silent prebuilt fallback then links without the ext-pump entrypoints the well-known routing already stripped stdlib features for — surfacing as undefined-`js_*` link errors two stages away from the cause.

**Fix:** `retain_workspace_declared_features()` filters `perry-runtime/*` / `perry-stdlib/*` cross-features against the checkout's declared features (features table + optional deps; fail-open when manifests are unreadable), applied before the build stamp so the stamp keys on what's actually built. Warns with what was dropped and why. The cargo-failure fallback message now states the likely consequence and remedy. Unit tests included.

## 2. `perry-ext-zlib` lacked the zstd surface
undici's web-fetch content decoding references `js_zlib_create_zstd_decompress` unconditionally. Only perry-stdlib's `compression` module carried the zstd codecs — and routing `node:zlib` to the ext archive strips that feature, so the link died on that one symbol.

**Fix:** port the full zstd surface into perry-ext-zlib: create factories, sync/async one-shots, and the streaming write-codec (`zstd::stream::write::{Encoder,Decoder}`), mirroring stdlib's ABI (declared in `runtime_decls`). Fallible zstd context allocation degrades to the same buffer-until-end path `createUnzip` uses.

## 3. `class X extends DOMException` didn't compile to something runnable
undici probes DOMException inheritability **at module load** (`websocketerror.js`: `class Test extends DOMException { … } new Test()`). DOMException was neither in codegen's builtin-parent list nor backed by a subclass initializer, so any binary with undici in its graph died at startup with `TypeError: DOMException is not a function`.

**Fix:** new runtime `js_dom_exception_subclass_init` (stamps `message`/`name`/`code` onto the subclass instance, defaults matching `js_dom_exception_new`), wired through both the explicit `super()` lowering and the implicit-ctor `NativeInstanceBase` chain walk. Verified with explicit-ctor and no-ctor repros.

## 4. panic-runtime dedup vs panic=abort stdlib
Prebuilt wrapper staticlibs are built `panic=unwind`; the auto-optimized stdlib is `panic=abort`. Two composing holes in `strip_dedup`:
- the nosharedeps name-containment rule never nominates the wrapper's `panic_unwind` member (an abort stdlib bundles `panic_abort` under a different member name), so the stale unwind copy survives;
- the localize pass hid the wrapper std-cgu's `__rust_drop_panic` definition while the sibling `panic_unwind` member still referenced it — and an abort stdlib has no replacement. Link dies on exactly that symbol.

**Fix:** nominate `panic_unwind` in the nosharedeps fixed-point (the existing protection logic keeps it exactly when the stdlib can't cover it — fail-safe), and skip localizing panic symbols a **sibling member** still references. Allocator shims deliberately stay always-localized: leaving the wrapper's system-malloc shim global would beat the runtime's mimalloc at link and break pointer classification (the known silent-console-loss failure mode) — we hit exactly that regression mid-fix and the comment documents it.

## Validation
- `cargo test -p perry --bin perry`: 753 passed, 0 failed (includes 2 new tests for the feature filter)
- `sfw` (enterprise) and `sfw-free`: compile → link → run, correct `--help` output, from their TS entrypoints with 28 npm packages in `perry.compilePackages`
- minimal DOMException-subclass repros (with and without explicit ctor) run correctly

Known follow-ups (not in this PR): `sfw-registry` still hits a startup `TypeError: Cannot read properties of undefined (reading 'test')`; `perry check --check-deps` doesn't route through the tsconfig-paths fallback that `perry compile` uses, so `#lib/*`-style subpath imports produce false R003 errors.

## Summary by CodeRabbit

* **New Features**
  * Added Node-compatible Zstandard (zstd) compression/decompression, including synchronous, asynchronous, and streaming APIs.
  * Added correct support for `class X extends DOMException` so `super(message, name)` initializes message, name, and error code properly.
  * Calling `RegExp(pattern, flags)` without `new` now correctly creates (or returns) a RegExp.

* **Bug Fixes**
  * Improved reliability of optimized native compilation by resolving multiple panic-mode and feature-selection edge cases.
  * Added clearer diagnostics when optimized compilation fails.

* **Tests**
  * Added coverage for zstd decoder streaming completion behavior.

### CODE REVIEW

**Comment by jdalton** [comment]
Follow-up resolved in this PR: the `sfw-registry` startup `TypeError: Cannot read properties of undefined (reading 'test')` traced to the globalThis `RegExp` sentinel — lodash's `runInContext` rebinds it (`var RegExp = context.RegExp`) and uses the **call form** to build `reIsNative`; the sentinel fell through to the noop thunk and returned `undefined`. Added `regexp_constructor_call_thunk` (ECMA-262 22.2.4 semantics incl. the `RegExp(re)` identity shortcut). All three Socket Firewall binaries (`sfw`, `sfw-free`, `sfw-registry`) now compile and run. 753/753 unit tests still pass.

File: .github/workflows/security-audit.yml [hunk 1]
```diff
@@ -64,6 +64,78 @@ jobs:
...
+          persist-credentials: false
+      - uses: astral-sh/setup-uv@c771a70e6277c0a99b617c7a806ffedaca235ff9 # v9.0.0
+      - name: Scan each skill
+        run: |
+          set -euo pipefail
+          for skill in .claude/skills/*/; do
+            echo "::group::skillspector ${skill}"
+            uv tool run --python 3.12 \
+              --from git+https://github.com/NVIDIA/skillspector@2eb84478 \
+              skillspector scan "${skill}" --no-llm
```
[the remark this answers is not in the document]
Comment (proggeramlug) [reply .github/workflows/security-audit.yml:135]: Implemented in d1fa9fa49. The workflow and external-tools.json now use the same full 40-character SkillSpector commit SHA: 2eb844780ab163f01468ecf142c40a2ec0fcaec0. The external-tool pin check and emitted uvx-command test pass.

File: .npmrc [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# npm-cli soak window (npm >= 11.17, DAYS) — governs the root npm install
+# that materializes the parity-test fixture deps (package-lock.json).
+# Managed by scripts/soak/soak.mts (`npm run soak` / `npm run soak:fix`).
+#
+# The pnpm-side soak (minimumReleaseAge, catalog) lives in
+# tools/pnpm-workspace.yaml, NOT here: pnpm only honors minimumReleaseAge
+# from a workspace yaml, and a workspace yaml at the repo root would mark
+# the npm-managed root as a pnpm workspace.
+min-release-age=7
```
[the remark this answers is not in the document]
Comment (proggeramlug) [reply .npmrc:9]: Implemented in d1fa9fa49. Root metadata now pins packageManager npm@12.0.1, requires npm >=11.17.0, and .npmrc enables engine-strict. A dry-run with pinned npm 12.0.1 is up to date; npm 11.2 is rejected with EBADENGINE before it can ignore min-release-age.

File: crates/perry-codegen/src/expr/this_super_call.rs [hunk 3]
```diff
@@ -709,6 +710,37 @@ pub(crate) fn lower(ctx: &mut FnCtx<'_>, expr: &Expr) -> Result<String> {
...
+                        );
+                        let current_class_name =
+                            ctx.class_stack.last().cloned().unwrap_or_default();
+                        crate::lower_call::apply_field_initializers_recursive(
+                            ctx,
+                            &current_class_name,
+                            crate::lower_call::FieldInitMode::SelfOnly,
+                        )?;
+                        return Ok(double_literal(f64::from_bits(crate::nanbox::TAG_UNDEFINED)));
+                    }
```
[the remark this answers is not in the document]
Comment (proggeramlug) [reply crates/perry-codegen/src/expr/this_super_call.rs:775]: Implemented in 781282f10. SuperCallSpread now initializes DOMException subclasses from the spread message/name arguments and applies SelfOnly fields. The compiled Perry regression covers both implicit and explicit spread constructors and passes.

File: crates/perry-runtime/src/event_target.rs [hunk 4]
```diff
@@ -326,6 +326,45 @@ pub extern "C" fn js_event_subclass_init(
...
+        exception,
+        b"message",
+        crate::value::js_nanbox_string(message_ptr as i64),
+    );
+    set_event_field(
+        exception,
+        b"name",
+        crate::value::js_nanbox_string(name_ptr as i64),
+    );
+    set_event_field(exception, b"code", dom_exception_code(&name_string));
```
[the remark this answers is not in the document]
Comment (proggeramlug) [reply crates/perry-runtime/src/event_target.rs]: Implemented in 781282f10. The DOMException initializer roots and reloads the receiver, arguments, and coerced strings; the shared field setter also roots values across allocations. Runtime/codegen checks and the compiled regression pass.

File: package.json [hunk 5]
```diff
@@ -1,5 +1,14 @@
...
   "type": "module",
+  "scripts": {
+    "soak": "node scripts/soak/soak.mts --check",
+    "soak:fix": "node scripts/soak/soak.mts --fix",
+    "deps:update": "node scripts/soak/update-deps.mts",
+    "tools:check": "node scripts/soak/external-tools.mts --check",
+    "tools:fix": "node scripts/soak/external-tools.mts --fix",
+    "tools:install": "node scripts/soak/external-tools.mts --install-all --shims",
+    "test:scripts": "node --test scripts/soak/*.test.mts"
+  },
```
[the remark this answers is not in the document]
Comment (proggeramlug) [reply package.json:11]: Implemented in d1fa9fa49. package.json declares the native type-stripping floor (^22.18.0 || >=23.6.0), and test:scripts passes a quoted glob for Node's test matcher instead of shell expansion. npm run test:scripts passes 49 tests with one platform-only skip.

File: crates/perry-runtime/src/object/global_this/ctor_thunks.rs [hunk 6]
```diff
@@ -25,6 +25,56 @@ pub(crate) extern "C" fn typed_array_constructor_call_thunk(
...
+        std::ptr::null_mut()
+    } else {
+        crate::builtins::js_string_coerce(pattern)
+    };
+    let flags_ptr = if flags_undefined {
+        std::ptr::null_mut()
+    } else {
+        crate::builtins::js_string_coerce(flags)
+    };
+    let re = crate::regex::js_regexp_new(pattern_ptr, flags_ptr);
```
[the remark this answers is not in the document]
Comment (proggeramlug) [reply crates/perry-runtime/src/object/global_this/ctor_thunks.rs]: Implemented in d1fa9fa49. RegExp call-form now roots both boxed inputs, roots the coerced pattern before coercing flags, roots the flags result, and reloads both string pointers only at js_regexp_new. cargo check -p perry-runtime --features regex-engine passes.
